platform/mellanox/mlnx-platform-api/sonic_platform/psu.py

```diff
@@ -12,44 +12,100 @@
+# but there are exceptions which will be handled by the following dictionary
+hwsku_dict_psu = {'ACS-MSN3700': 1, 'ACS-MSN3700C': 1, 'ACS-MSN3800': 1}
+psu_profile_list = [
+    # default filename convention
+    {
+        PSU_CURRENT : "power/psu{}_curr",
+        PSU_VOLTAGE : "power/psu{}_volt",
+        PSU_POWER : "power/psu{}_power"
+    },
+    # for 3700, 3700c, 3800
+    {
+        PSU_CURRENT : "power/psu{}_curr",
+        PSU_VOLTAGE : "power/psu{}_volt_out2",
+        PSU_POWER : "power/psu{}_power"
+    }
+]
+
 class Psu(PsuBase):
     """Platform-specific Psu class"""
-    def __init__(self, psu_index):
+    def __init__(self, psu_index, sku):
         global psu_list
         PsuBase.__init__(self)
         # PSU is 1-based on Mellanox platform
         self.index = psu_index + 1
         psu_list.append(self.index)
-        self.psu_path = "/var/run/hw-management/thermal/"
-        self.psu_oper_status = "psu{}_pwr_status".format(self.index)
-        self.psu_presence = "psu{}_status".format(self.index)
-        if os.path.exists(os.path.join(self.psu_path, self.psu_presence)):
-            self.presence_file_exists = True
+        self.psu_path = "/var/run/hw-management/"
+        self.psu_oper_status = "thermal/psu{}_pwr_status".format(self.index)
+
+        if sku in hwsku_dict_psu:
+            filemap = psu_profile_list[hwsku_dict_psu[sku]]
+        else:
+            filemap = psu_profile_list[0]
+
+        print "sku {} filemap {}".format(sku, filemap)
+        psu_voltage = filemap[PSU_VOLTAGE].format(self.index)
+        if os.path.exists(os.path.join(self.psu_path, psu_voltage)):
+            self.psu_voltage = psu_voltage
+        else:
+            self.psu_voltage = None
+        psu_curr = filemap[PSU_CURRENT].format(self.index)
+        if os.path.exists(os.path.join(self.psu_path, psu_curr)):
+            self.psu_current = psu_curr
+        else:
+            self.psu_current = None
+        psu_power = filemap[PSU_POWER].format(self.index)
+        if os.path.exists(os.path.join(self.psu_path, psu_power)):
+            self.psu_power = psu_power
+        else:
+            self.psu_power = None
+        psu_presence = "thermal/psu{}_status".format(self.index)
+        if os.path.exists(os.path.join(self.psu_path, psu_presence)):
+            self.psu_presence = psu_presence
         else:
-            self.presence_file_exists = False
+            self.psu_presence = None
         fan = Fan(psu_index, psu_index, True)
         if fan.get_presence():
             self._fan = fan
 
-    def get_status(self):
+    def _read_generic_file(self, filename, len):
+        """
+        Read a generic file, returns the contents of the file
+        """
+        result = 0
+        try:
+            with open(filename, 'r') as fileobj:
+                result = int(fileobj.read())
+        except:
+            logger.log_info("Fail to read file {}, maybe it doesn't exist".format(filename))
```

Comment:
rephase, we should give clear information in the log.

Reply:
fixed.